packages/astro/src/i18n/middleware.ts

```diff
@@ -1,18 +1,31 @@
 import { appendForwardSlash, joinPaths } from '@astrojs/internal-helpers/path';
-import type { MiddlewareEndpointHandler, RouteData, SSRManifest } from '../@types/astro.js';
+import type {
+	Locales,
+	MiddlewareEndpointHandler,
+	RouteData,
+	SSRManifest,
+} from '../@types/astro.js';
 import type { PipelineHookFunction } from '../core/pipeline.js';
+import { getPathByLocale, normalizeTheLocale } from './index.js';
 
 const routeDataSymbol = Symbol.for('astro.routeData');
 
-// Checks if the pathname doesn't have any locale, exception for the defaultLocale, which is ignored on purpose
-function checkIsLocaleFree(pathname: string, locales: string[]): boolean {
-	for (const locale of locales) {
-		if (pathname.includes(`/${locale}`)) {
-			return false;
+// Checks if the pathname doesn't have any locale, exception for the defaultLocale, which is ignored on purpose.
```

Comment:
Nit: Could we reformulate it like this, since it has the opposite result of the function it replaces?
```suggestion
// Checks if the pathname has any locale, exception for the defaultLocale, which is ignored on purpose.
```